### DISCUSSION TITLE
How would you  optimize discrete random variable parameters using an external error function in pgmpy?

### DISCUSSION THREAD

**Discussion by ferencbartok** [pgmpy/pgmpy#1750] [Q&A]
I have a Bayesian network with discrete random variables. The initial probabilities and conditional probabilities for the variables are set by experts. There are external requirements that tell us based on specific sets of evidence (not for all variables) what the inferred probabilities should be for a set of nodes. I’d like to optimize the variable parameters so that as many requirements are satisfied as possible.

Example: you have A, B, C, D network, where A is the parent of B and C. B and C are the parents of D. Let’s say there is a requirement that if A=1 then D should be 1 with about 65%. I’d like to calculate the parameters of B, C (and sometimes D) so that D=1 is around 65%.

In this [example](https://pgmpy.org/examples/Learning%20Parameters%20in%20Discrete%20Bayesian%20Networks.html) parameter learning is done via generating data. How could I incorporate an error function and minimize that during parameter learning? Do I have to generate data with that in mind or would there be another way you think?

The error function would look something like this:

- do inference based on evidence for the nodes in question
- calculate some sort of distance between the expected value and the calculated inference

Any ideas?

**Comment by ankurankan** [comment]
@ferencbartok This sounds a lot like iterative proportional fitting to me. Could you have a look at the problem statement described in: https://arxiv.org/pdf/1207.1356 (the first paragraph of the introduction describes the problem setting). If that is exactly your usecase, I have a local implementation of the algorithm described in the paper that I can share or add to pgmpy.

**Reply by ferencbartok** [reply]
Thanks for the quick reply! It seems like it's very close to my problem. The only difference I see is that I think in the linked problem statement there is no mention of evidence. We have probability constraints given some evidence. We also have requirements where there are 0 evidence and just the probability constraints, then I think it is exactly the same problem as you linked. Would be nice to somehow create a solution for both. I'd be happy if you added that solution to pgmpy or shared the code.

What do you think we could do for the case where we have to set some evidence and then try to satisfy the probability constraints?

Side question: do you think something like this would work using for example scipy's optimizer?:
```
def objective_function(parameters):
    update_bayesian_network(parameters)
    calculated_probabilities = perform_inference(evidence)
    distance = calculate_distance(expected_probabilities, calculated_probabilities) # basically the probability constraints

    return distance


def test(network: list[Node]):
    solver = PgmpySolver()
    solver.initialize(network)
    initial_flattened_cpts = flatten_cpts(network)
    result = minimize(objective_function, initial_flattened_cpts, method='L-BFGS-B')
```

This sort of brute-force solution that I'm planning to try first, but may not even work. Of course, some more sophisticated approach would be much better like the one you linked, or using pgmpy's parameter learning in a smart way.

**Reply by ankurankan** [reply]
@ferencbartok 
> What do you think we could do for the case where we have to set some evidence and then try to satisfy the probability constraints?

I think the IPFP algorithm should work in that case as well. The intermediate step in the standard algorithm computes the marginal distributions on the constrained variables, I think we should be able to replace that by conditional distributions (i.e. the evidence usecase that you have described). I am not sure about the convergence of the algorithm in that case though. I will try to clean up my implementation and figure out a way to add it to pgmpy.

> Side question: do you think something like this would work using for example scipy's optimizer?:

I think this optimization approach should work as well. Previously, I have written similar optimizers for optimizing functions of posterior distributions. I am assuming that you want minimal changes to your expert model such that it satisfies your constraints (this is what the IPFP algorithm does as well)? In that case, you would need to incorporate that as well in your loss/distance function. I would also suggeset trying out non-gradient based estimators (such as differential evolution), I am not sure how gradient based optimization would work on these discrete variables.

**Reply by ferencbartok** [reply]
> I will try to clean up my implementation and figure out a way to add it to pgmpy.

Thank you!

> I am assuming that you want minimal changes to your expert model such that it satisfies your constraints (this is what the IPFP algorithm does as well)? In that case, you would need to incorporate that as well in your loss/distance function. I would also suggeset trying out non-gradient based estimators (such as differential evolution), I am not sure how gradient based optimization would work on these discrete variables.

For now, it is not important to be as close as possible to the expert model. I've read in the paper that IPFP does this, which can be nice but not a hard requirement as of now in our use case. 
Thanks for the suggestion of other estimators! Will look into those as well. For now I could setup a simple minimization using scipy and infer which works correctly for a very small network. Will try a much bigger network soon and we'll see how it does. :)
